mountpoint-s3-fuser/src/session.rs

```
@@ -148,8 +148,35 @@ impl<FS: Filesystem> Session<FS> {
     /// calls into the filesystem.
     /// This version also notifies callers of kernel requests before and after they
     /// are dispatched to the filesystem.
-    pub fn run_with_callbacks<FA, FB>(&self, mut before_dispatch: FB, mut after_dispatch: FA) -> io::Result<()> 
-    where 
+    pub fn run_with_callbacks<FA, FB>(&self, before_dispatch: FB, after_dispatch: FA) -> io::Result<()>
+    where
+        FB: FnMut(&Request<'_>),
+        FA: FnMut(&Request<'_>),
+    {
+        // Check if UNSTABLE_MOUNTPOINT_CLONE_FUSE_FD environment variable is set to true
+        // If not set or not "true", don't clone the channel (default behavior)
+        let clone_fd = std::env::var("UNSTABLE_MOUNTPOINT_CLONE_FUSE_FD")
+            .map(|val| val.to_lowercase() == "true")
+            .unwrap_or(false);
+
+        info!("FUSE channel cloning: {}", if clone_fd { "enabled" } else { "disabled" });
+
+        // Create a worker channel for this thread if cloning is enabled
+        // Otherwise use the original channel directly
+        if clone_fd {
+            // Create a worker channel for this thread using FUSE_DEV_IOC_CLONE
+            // This allows multiple threads to read from the FUSE device without contention
+            let worker_channel = self.ch.clone_channel()?;
```

Should we think about falling back to the default on error?

---

yeah; I nearly wrote that code and then decided not to.  My reasoning is that the customer opted into this behaviour by setting the flag, and it'd be better to fail rather than invisibly falling back.  Especially in e.g. our performance tests, we need to know that the configuration was actually active.

If this were default behaviour we'd definitely have to fall back – the flag has been around for a while, introduced in kernel 4.2 I think, but not sure if any of our supported distros have kernels with it, for example.